### PR TITLE
[Prim][PIR] Support dynamic shape for add_grad

### PR DESCRIPTION
**Pull request by zeroRains**
### PR Category

Operator Mechanism

### PR Types

New features

### Description

前置PR：https://github.com/PaddlePaddle/Paddle/pull/65002

为add_grad添加动态shape支持

### CODE REVIEW

File: test/prim/pir_prim/test_prim_sub_graph_backward_dynamic_shape.py [hunk 1]
```diff
@@ -206,5 +210,147 @@ def setUp(self):
...
+        for ref, actual in zip(res_ref, res):
+            np.testing.assert_allclose(
+                ref, actual, rtol=self.tol, atol=self.tol
+            )
+
+        for dr, d in zip(grad_ref, grad):
+            np.testing.assert_allclose(dr, d, rtol=self.tol, atol=self.tol)
+
+
+class TestPrimAddWithGrad1(TestPrimTwoWithGrad):
```
Comment (cyber-pioneer) [inline test/prim/pir_prim/test_prim_sub_graph_backward_dynamic_shape.py:265]: 补充一个前向自动broadcast的单侧

Comment (cyber-pioneer) [reply test/prim/pir_prim/test_prim_sub_graph_backward_dynamic_shape.py:265]: 不同rank的

Comment (zeroRains) [reply test/prim/pir_prim/test_prim_sub_graph_backward_dynamic_shape.py:265]: Done

**Review by cyber-pioneer — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Pull request merged**